src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+use crate::types::PyString;
+#[cfg(not(windows))]
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+impl ToPyObject for OsStr {
+    fn to_object(&self, py: Python) -> PyObject {
+        // If the string is UTF-8, take the quick and easy shortcut
+        if let Some(valid_utf8_path) = self.to_str() {
+            return PyString::new(py, valid_utf8_path).into();
+        }
+
+        // Take advantage of Python's surrogateescape encoding to support arbitrary byte
+        // sequences instead of limiting ourselves to UTF-8
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is this the appropriate function? Does it use the surrogateescape error
+                // handler? Should it be PyUnicode_DecodeLocaleAndSize instead maybe?
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
```

Comment:
Good question - I think probably `DecodeLocaleAndSize` is the right choice 👍

Reply:
I'm curious why? I'd tentatively guess that PyUnicode_DecodeFSDefaultAndSize is a special case of PyUnicode_DecodeLocaleAndSize, but it chooses the error handler appropriate for file paths; the same one that Python itself uses. The Python docs don't say much about specifics, unfortunately

Reply:
I checked that `interp->fs_codec.error_handler` is used in `PyUnicode_DecodeFSDefaultAndSize`, but how is it different from other handlers?

Reply:
The two error handlers I know of, and are mentioned in the Python docs, are "strict" and "surrogateescape". "strict" will error out if invalid Unicode data is encountered. "surrogateescape" will handle any invalid bytes by encoding them in the Low Surrogates Unicode block.

Therefore, "surrogateescape" can decode any arbitrary bytes losslessly by parsing UTF-8 where it can, and saving the rest as surrogates instead. Hence "surrogateescape" is well-suited for parsing filenames, since Unix filenames _are_ actually arbitrary bytes.

I did an experiment:
```sh
➜  ~ touch $(echo -e "\xFA\xFB\xFC\xFD\xFE\xFF") # invalid UTF-8
➜  ~ python3
>>> import os; os.listdir()
['\udcfa\udcfb\udcfc\udcfd\udcfe\udcff']
```
As you can see, the invalid bytes I put into the filename are being encoded in the Unicode `0xdc__` block, which strongly suggests that Python uses "surrogateescape" as its file system string parser.

All this complexity is encapsulated by using PyUnicode_DecodeFSDefaultAndSize, which is specifically made for filesystem operations and does _the right thing™_. Hence I tend to prefer PyUnicode_DecodeFSDefaultAndSize over PyUnicode_DecodeLocaleAndSize

Reply:
Thank you for your explanation, but isn't it look like the handler is not always `surrogateescape`?
https://github.com/python/cpython/blob/v3.9.1/Python/initconfig.c#L1695

Anyway, I agree that `DecodeFSDefault` is better. There's no merit in choosing the error handler by ourselves.

Reply:
Heh, seems like in reality it's even more complex than what I described above. It applies a third error handler for "legacy Windows". I guess this makes it even clearer that it's probably better to leave the error handler decision to Python

Reply:
Yes I agree let's go with `DecodeFSDefault` - I'd mistakenly understood previously that we want `surrogateescape` in all circumstances.